miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,802 @@
+use alloc::vec::Vec;
+use core::{
+    iter::{self, FusedIterator},
+    num::NonZero,
+};
+
+use winter_utils::{Deserializable, DeserializationError, Serializable};
+
+use super::{
+    EmptySubtreeRoots, MerkleError, MerklePath, RpoDigest, SMT_MAX_DEPTH, ValuePath, Word,
+};
+
+/// A different representation of [`MerklePath`] designed for memory efficiency for Merkle paths
+/// with empty nodes.
+///
+/// Empty nodes in the path are stored only as their position, represented with a bitmask. A
+/// maximum of 64 nodes in the path can be empty. The number of empty nodes has no effect on memory
+/// usage by this struct, but will incur overhead during iteration or conversion to a
+/// [`MerklePath`], for each empty node.
+///
+/// NOTE: This type assumes that Merkle paths always span from the root of the tree to a leaf.
+/// Partial paths are not supported.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct SparseMerklePath {
+    /// A bitmask representing empty nodes. The set bit corresponds to the depth of an empty node.
+    empty_nodes: u64,
+    /// The non-empty nodes, stored in depth-order, but not contiguous across depth.
+    nodes: Vec<RpoDigest>,
+}
+
+impl SparseMerklePath {
+    /// Constructs a sparse Merkle path from an iterator over Merkle nodes that also knows its
+    /// exact size (such as iterators created with [Vec::into_iter]). The iterator must be in order
+    /// of deepest to shallowest.
+    ///
+    /// Knowing the size is necessary to calculate the depth of the tree, which is needed to detect
+    /// which nodes are empty nodes. If you know the size but your iterator type is not
+    /// [ExactSizeIterator], use [`SparseMerklePath::from_iter_with_depth()`].
+    pub fn from_sized_iter<I>(iterator: I) -> Result<Self, MerkleError>
```

Comment:
Errors should be documented? https://github.com/0xPolygonMiden/crypto/pull/389#discussion_r2009910558